src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -621,3 +645,385 @@ def _find_basic_blocks(self, nodelist):
+                        chains.append(chain)
+
+                # Check if there is an ancestor Assignment.
+                ancestor = self._reference.ancestor(Assignment)
+                if ancestor is not None:
+                    # If the reference is not the lhs then we can ignore
+                    # the RHS.
+                    if ancestor.lhs is self._reference:
+                        end = ancestor.walk(Node)[-1]
+                        # Add the rhs as a potential basic block with
+                        # different start and stop positions.
+                        chain = DefinitionUseChain(
+                            self._reference.copy(),
+                            ancestor.rhs.children[:],
+                            start_point=ancestor.rhs.abs_position,
+                            stop_point=end.abs_position,
+                        )
+                        control_flow_nodes.append(None)
+                        chains.append(chain)
+                        # N.B. For now this assumes that for an expression
+                        # b = a * a, that next_access to the first Reference
+                        # to a should not return the second Reference to a.
+
+            # For backwards we want to reverse the order.
+            chains.reverse()
+            control_flow_nodes.reverse()
+            for i, chain in enumerate(chains):
+                # Compute the defsout, killed and reaches for the block.
+                chain.find_backward_accesses()
+                cfn = control_flow_nodes[i]
+
+                if cfn is None:
+                    # We're outside a control flow region, updating the reaches
+                    # here is to find all the reached nodes.
+                    for ref in chain._reaches:
+                        # Add unique references to reaches. Since we're not
+                        # in a control flow region, we can't have added
+                        # these references into the reaches array yet so
+                        # they're guaranteed to be unique.
+                        found = False
+                        for ref2 in self._reaches:
+                            if ref is ref2:
+                                found = True
+                                break
+                        if not found:
+                            self._reaches.append(ref)
+                    # If we have a defsout in the chain then we can stop as we
+                    # will never get past the write as its not conditional.
+                    if len(chain.defsout) > 0:
+                        # Reset the start and stop points before returning
+                        # the result.
+                        self._start_point = save_start_position
+                        self._stop_point = save_stop_position
+                        return self._reaches
+                else:
+                    # We assume that the control flow here could not be taken,
```

Comment:
Again, maybe its just me but "could not be taken" seems that something when wrong, so maybe "could be 'not taken'" to make clear that we talk about the chain branches?

Reply:
Yeah, changed, that makes more sense.